sync_worker.go

```diff
@@ -0,0 +1,224 @@
+package main
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"golang.org/x/crypto/nacl/box"
+)
+
+var (
+	errBecameLeader    = errors.New("became enclave leader")
+	errNonceRequired   = errors.New("nonce is required")
+	errInProgress      = errors.New("key sync already in progress")
+	errInvalidNonceLen = errors.New("invalid nonce length")
+	errDecrypting      = errors.New("error decrypting enclave keys")
+)
+
+// workerSync holds the state and code that we need for a one-off sync with a
+// leader enclave.  workerSync implements the http.Handler interface because the
+// sync protocol requires two endpoints on the worker.
+type workerSync struct {
+	attester
+	installKeys   func(*enclaveKeys) error
+	ephemeralKeys chan *boxKey
+	nonce         chan nonce
+	becameLeader  chan struct{}
+}
+
+// asWorker returns a new workerSync object.
+func asWorker(
+	installKeys func(*enclaveKeys) error,
+	becameLeader chan struct{},
+) *workerSync {
+	return &workerSync{
+		attester:      &dummyAttester{},
+		installKeys:   installKeys,
+		becameLeader:  becameLeader,
```

Comment:
Could a worker maliciously claim leader status causing confusion or running a separate image that allows the key material to be exfiltrated?

I didn't see details of this aspect of the protocol in the document about leader changes. Could we document that as well?

Reply:
The leader only syncs with workers that run identical code, i.e., the worker's image ID in the attestation document is identical to the leader's. On top of that, workers cannot claim leader status because the leader is reachable via a separate domain.

I added a clarifying sentence in 0f42e6b.